### ISSUE TITLE
Add common interface to classes

### ISSUE THREAD

**Issue by stamanker** [rubenlagus/TelegramBots#593]
Hi!
Could you please add & implement interface for classes that has method 'setReplyMarkup': 
EditMessageText, SendMessage, SendPhoto, EditMessageMedia.
Or how can I call that method for any of these classes without knowing exactly which exactly? 
Class 'BotApiMethod' doesnt have that method.

Thanks )

**Comment by Chase22** [comment]
All ApiMethod classes are direct mappings of the Api methods from the telegram docs.
They don't have a common interface because they are separate methods and adding interfaces for all parameters those methods have in common will get messy.

It's probably better to build a wrapper yourself or figure out if you need to access the methods generic

**Comment by stamanker** [comment]
Yes, it may be solved with wrapper around for _every_ of these classes. 
But if they would have implemented common interface with that method, then all of them could be used via reference of this interface without any harm.

**Comment by Chase22** [comment]
If we add an interface for setReplyMarkup why not for setChatId or formatMode or file id or message?

What exactly is it your trying to do? Maybe there's a good way for that without implementing interfaces

**Comment by stamanker** [comment]
Well, yes, actually all these methods have to be in common interface(s) as well )
I'm tring to implement common behavior for all these instances. For example, I want to add buttons for edited post. Post may be text or media. In one case i need to use 'EditMessageText' in other - 'EditMessageMedia'.

**Comment by aNNiMON** [comment]
I also encountered the inconvenience of using some methods, so I made [a wrapper](https://github.com/aNNiMON/tgbots-module/blob/master/module/src/main/java/com/annimon/tgbotsmodule/api/methods/). Take a look to `interfaces` package.
You can take this code as the basis for your wrapper implementation.

**Comment by stamanker** [comment]
@Chase22 could you plz explain why you don't want to add simple interface?

**Comment by Chase22** [comment]
@stamanker We can start out that ApiMethods already have a common base class `BotApiMethod<T>` which is an abstract class. So implementing a common interface for just some BotMethods is already impossible.

So we'd have to use an abstract class. Making the inheritance already 4 levels deep. And to cover every case we will have to add more layers of abstract classes. Basically 1 for every common methods. Then the api methods have to extend them in the proper order to only get the methods they need.

It would make the entire libary uselessly complex and unflexible. What happens if a ApiMethod changes and they add a method? We'll have to possibly change around a lot of inheritance to cover that.

And the last point: It is not correct. Interfaces should be used to abstract classes on a level where you don't care how something is implemented, just that is it implemented. Interfaces should rarely extend other interfaces and abstract classes should rarely extend other abstract classes. 

The bot methods don't have a common interface, because they don't do common things. They just happen to have a few common parameters.

**Comment by stamanker** [comment]
@Chase22 ,
Ok, let's go from practice side:
```if (message.hasVideo()) {
                    EditMessageMedia editObject = new EditMessageMedia().setChatId(chatId).setMessageId(messageId);
                    editObject.setMedia(
                            new InputMediaVideo().setMedia(message.getVideo().getFileId())
                    );
                    setButtons(data, editObject);
                    execute(editObject);
                } else {
                    EditMessageText editObject = new EditMessageText()
                            .setMessageId(messageId)
                            .setChatId(chatId)
                            .setText(message.getText());
                    setButtons(data, editObject);
                    execute(editObject);
                }```
How can I generalize code without excessive wrappers to call `execute` only once for both cases?
The same for setButtons: 
```
private void setButtons(MsgData data, Object event) {
         ...
        if (event instanceof EditMessageText) { 
            ((EditMessageText) event).setReplyMarkup(replyMarkup);
        } else if (event instanceof EditMessageMedia) {
            ((EditMessageMedia) event).setReplyMarkup(replyMarkup);
        }
        ...
```
May be I missed something and this could be done without `instanceof` ? Please, advise :)
Thanks

**Comment by Chase22** [comment]
@stamanker 
```java
if (message.hasVideo()) {
  EditMessageMedia editObject = new EditMessageMedia()
    .setChatId(chatId)
    .setMessageId(messageId)
    .setMedia(
        new InputMediaVideo().setMedia(message.getVideo().getFileId())
    ).setReplyMarkup(
      getMarkup(data)
    );
    execute(editObject);
} else {
  EditMessageText editObject = new EditMessageText()
    .setMessageId(messageId)
    .setChatId(chatId)
    .setText(message.getText())
    .setReplyMarkup(
      getMarkup(data)
    );
    execute(editObject);
}
```

**Comment by stamanker** [comment]
@Chase22 
This makes no practical difference, the code is duplicated anyway.
And especially this is obvious for `execute`.
And thats how it could look like:
```
SomeInterfaceForAllTheseEvents editObject;
if (message.hasVideo()) {
    editObject = new EditMessageMedia().setMedia(
        new InputMediaVideo().setMedia(message.getVideo().getFileId())
    );
} else {
    editObject = new EditMessageText().setText(message.getText());
}
editObject.setChatId(chatId).setMessageId(messageId)
setButtons(data, editObject);
execute(editObject);
```
No wrappers, no duplicated code, just pure polymorphysm...

**Comment by Chase22** [comment]
@stamanker i'd love to see you make a working prototype for a common abstract class (interfaces don't work thanks to BotApiMethod) that doesn't blow up the libary nedlessly

**Comment by rubenlagus** [comment]
Adding the interface for those classes is not an actual problem (I can agree in adding them) but, in any case, you will still need to do a casting from BotApiMethod to the new interface since "execute" method is not accepting that kind of interfaces directly.

Feel free to do a pull request with the change and we can see how it fits.

**Comment by SrMouraSilva** [comment]
There is any evolution in this way?
